[DEVEXP-457] Create app from source

jcantrill opened the pull request:
@liggitt @jwforres Please do an initial review with regards to structure and what I'm pushed for filters, directives, specs, routes to see if I am moving in the direction you would expect

jcantrill commented:
@jwforres This latest commit wires the 'create' action to the back-end.  Still need to add some validations around validating labels, validating a user entering appname, warning a user if a generated appname is taken.

jcantrill commented:
@jwforres Added validations and fixes to post resources to the server that will actually trigger a build.  I think its ready to be evaluated by a wider audience.  Also, I think we should decide if the way I implemented the routing functionality is what we really desire and if the UI makes sense for routing. The styling could also use another once over.
Maybe @pweil- should have a look to see its reasonable to him as well.
cc @bparees @csrwng @EmilyDirsh @sg00dwin

jwforres commented:
See comments in below screenshots:

![create_route_for_me](https://cloud.githubusercontent.com/assets/1467629/6667185/d91227d4-cbc0-11e4-8f2d-e73acdf8c14a.png)
![secure_route_config](https://cloud.githubusercontent.com/assets/1467629/6667186/d915a5da-cbc0-11e4-917b-cedb4c5f5c1d.png)

Also, not sure i like the "Specify route" label in combination with "Create a route for me".  Seems like specify should be more like "Let me specify a route" or "I have a route already"

liggitt commented:
Can we combine the "create for me" and "specify" cases, and just put a default value in the route text box?

Also, instead of a checkbox for secure/insecure, can we put a combo in front of the text field that lets you select "http://" or "https://"? I think that would clarify a few things: 1) what the option does to the route, 2) that you don't need the scheme in the route textbox

liggitt commented:
Also, should we have the cert/key/ca be selected as a file browse instead of pasting in content?

jwforres commented:
@jcantrill where do you get the auto-generated route from, is that something you have hard-coded?

Should check with @pweil- and @smarterclayton related to the skydns stuff @smarterclayton added.  We probably need to send another config back to the web console in config.js for the subdomain that routes will be auto-generated at.

pweil- commented:
/cc @ramr Ram, it looks like this could use a hook into your route allocator code merged in https://github.com/openshift/origin/pull/1291 for auto-generating route names

jwforres commented:
![scaling](https://cloud.githubusercontent.com/assets/1467629/6667488/c4ae40a0-cbc2-11e4-8f5b-7c8af9809ef8.png)

See comment in screenshot for style fix on scaling input.   Also should we be letting them specify 0 for scaling, right now you have min set to 1, but I could see someone wanting to get everything set up with 0 scale to start with.

jwforres commented:
"Automatically build new images when code changes" is a deceptive label, we aren't doing that level of magic for them.  Need to be more clear about what exactly we are giving them (which are webhooks they can then use).

jwforres commented:
Possibly instead of "Autodeploy when a new image is available "   make it "Autodeploy when a new builder image is available"   so it is clear what image we are talking about?

jwforres commented:
Don't forget to update the help tip for scaling.  Say something about it controlling the number of running instances of your built image.

jwforres commented:
nothing tells me what the "Name" is actually being used for, I may want to know so that it helps me decide what to call it

csrwng commented on assets/app/index.html:
```diff
@@ -124,14 +127,21 @@
+        <script src="scripts/controllers/catalog/images.js"></script>
+        <script src="scripts/controllers/create.js"></script>
         <script src="scripts/directives/date.js"></script>
+        <script src="scripts/directives/oscFormSection.js"></script>
```
is there a reason to prefix these with 'osc' (I would think all directives are part of the openshift console)?

jwforres replied:
this was the discussion we had the other day about prefixing our directives, its apparently a "best practice" to prefix them, so <osc-form-section>.   It is debatable whether the file name really needs to have the prefix since in some of these files we are putting multiple directives.

csrwng replied:
If we do decide to prefix the files, it'd make sense to eventually prefix all of them

csrwng commented on assets/app/scripts/controllers/catalog/images.js:
```diff
@@ -0,0 +1,72 @@
+      console.log("project image repos", $scope.projectImageRepos);
+    });
+
+    DataService.list("imageRepositories", {namespace: "openshift"}, function(imageRepos) {
```
May be good to have a service that returns the default namespace ($openshiftNamespace) so it's in one place

bparees commented:
sorry, jumping in late, but it looks like we're missing an option for "rebuild when builder image changes" or did we decide that's an expert mode configuration and we'll just default it to true here?

bparees commented:
also we're not doing any validation on the source-url input field?  (at least confirming it's a valid url?)

jcantrill commented:
@bparees Correct there is no source url validation.  @jwforres ???  In regards to rebuild when builder image changes is that "Autodeploy when a new builder image is available"?

bparees commented:
@jcantrill no, and in fact i think you've got mixed wording there.  will comment directly on that code.

but in any case deploy is different from build.

bparees commented on assets/app/views/create/fromimage.html:
```diff
@@ -0,0 +1,287 @@
+                  <div>
+                    <strong>Autodeploy when deployment configuration changes: </strong>
+                    <span>{{deploymentConfig.deployOnConfigChange | yesNo}}</span>
+                  </div>
```
assuming this is ties to line 152, this line should not say "builder", it's just autodeploy when a new image is available.

what this should translate into is "when a new version of my application image is created, either because i manually pushed something, or a build pushed something, redeploy my app"

That's separate from what I think is still missing here which is "if my builder image (eg ruby-20-centos7) changed, rebuild my application".  That rebuild might then trigger a redeploy, but that's a separate issue.

jcantrill commented:
@jwforres @bparees @csrwng @sg00dwin Rebased to latest and updated to address previous comments(routing, key/value wigit for labels and env vars).  It really needs some styling help especially in regards to the file upload and routing stuff

jcantrill commented:
@jwforres @bparees @sg00dwin Updated PR to add imageChange trigger

bparees commented:
imagechange bits look right.

jcantrill commented:
@jwforres @bparees @sg00dwin  @pweil-  Rebased and updated based on steve's style changes and the routing changes requested.  Added in Label value validations.  Tested that it creates resources, though I'm not completely convinced the way I'm generating routes is correct.  Nor do I really know how to go about testing them so I am unable to confirm they actually work correctly.

bparees commented:
@jcantrill and you're just creating a service/optional route for the first exposed port in the image, right?  or did you end up doing something additional after this afternoon's discussion?

jcantrill commented:
@bparees I modified the app generator to always create 2 services, one on the first exposed port, the other on 443.  Additionally, I modified it to create at most 2 routes based on the choice of the user.  After reworking the UI to support the most recent changes, IMO I think it would be a more consistent experience to  make it more like v2 where you get a singular, optional, public entry point with minimal or no configuration and require them to drop down to the CLI if they want something different.

smarterclayton commented:
Why are we creating two services?  We explicitly said we would only create the first for new-app...  Multiport services are coming very soon


> On Apr 2, 2015, at 7:07 AM, Jeff Cantrill <notifications@github.com> wrote:
> 
> @bparees I modified the app generator to always create 2 services, one on the first exposed port, the other on 443. Additionally, I modified it to create at most 2 routes based on the choice of the user. After reworking the UI to support the most recent changes, IMO I think it would be a more consistent experience to make it more like v2 where you get a singular, optional, public entry point with minimal or no configuration and require them to drop down to the CLI if they want something different.
> 
> —
> Reply to this email directly or view it on GitHub.
>

bparees commented:
regardless of impending multiport services... always creating something on 443 seems like a mistake, why would we assume every image listens on 443?

smarterclayton commented:
If someone tells us it has one port or more, pick the first and expose it.  We can eventually shift to headless services for those that don't have a port.


> On Apr 2, 2015, at 10:00 AM, Ben Parees <notifications@github.com> wrote:
> 
> regardless of impending multiport services... always creating something on 443 seems like a mistake, why would we assume every image listens on 443?
> 
> —
> Reply to this email directly or view it on GitHub.
>

bparees commented:
@smarterclayton what's the point in even having a service for something with no port?

smarterclayton commented:
Spreading, grouping, readiness assessment, etc.  there's a good description in a Kube issue upstream, search for "headless services".


> On Apr 2, 2015, at 3:40 PM, Ben Parees <notifications@github.com> wrote:
> 
> @smarterclayton what's the point in even having a service for something with no port?
> 
> —
> Reply to this email directly or view it on GitHub.
>

bparees commented:
@jcantrill we need to find a way to get this landed for beta3 so if we have to limit it to one service using the lowest exposed port for now, let's do that and come back around to the rest of this.  People can always define more services+routes by hand.  (eg for passthrough routing)

However if it's feasible, the approach i'd prefer to see is they get a list of the ports the image exposes and for each one they get a drop down where they can choose "do not expose, expose as service, expose as route, expose as custom route" (the same options we were already offering).

And for bonus points, the ability to click a "+" to add additional ports to be exposed.

But again, we can iterate on this.

smarterclayton commented:
> On Apr 2, 2015, at 4:27 PM, Ben Parees <notifications@github.com> wrote:
> 
> @jcantrill we need to find a way to get this landed for beta3 so if we have to limit it to one service using the lowest exposed port for now, let's do that and come back around to the rest of this. People can always define more services+routes by hand. (eg for passthrough routing)
> 
> However if it's feasible, the approach i'd prefer to see is they get a list of the ports the image exposes and for each one they get a drop down where they can choose "do not expose, expose as service, expose as route, expose as custom route" (the same options we were already offering).
> 
No disagreement, but we need that after create more than before create, so I don't want to require people to delete and recreate services to get an outcome like it.  More prioritizing the overall story rather than getting focused on one loop.
> And for bonus points, the ability to click a "+" to add additional ports to be exposed.
> 
> But again, we can iterate on this.
> 
> —
> Reply to this email directly or view it on GitHub.
>

jcantrill commented:
@bparees @smarterclayton I would prefer to revert our route choices to what we had prior to this latest version of the PR where you could choose to create at most 1 route and that would be routed to the only service that was created.  I still don't understand why we decided to allow you to have up to 2 routes; that seems like a 'drop to the CLI' usecase.  I thought the second service was required based on my conversations with @pweil-, but maybe my whole understanding of routing is misguided.  IMO, we should revert to allowing a single route and creating a single service that exposes the 'first' port.  All other usecases should either be CLI or follow-on functionality.

bparees commented:
@jcantrill agreed (follow-on is more likely than just punting to the CLI entirely, though, imho).

@smarterclayton didn't follow your comments about delete+recreating services.  I mean i agree we don't want people to have to do that, but i don't see anything in this discussion that implies you would do so.  We're just talking about having users create additional services/routes outside the web console if they need something the web console flow doesn't cover.

smarterclayton commented:
Complexity in a create flow is less important than post creation complexity.  You are more likely to create something simple and then tune it than you are to create something complex whole cloth.

bparees commented:
@smarterclayton :

Saying "complexity in a create flow is less important" could be interpreted as "it's ok to be super complex in the create flow, it's not a problem, so let's make it super complex" or "it's not necessary to include complexity in the create flow".  I'm assuming you meant the latter, so let me attempt to paraphrase:  "it's less critical we provide all the possible options during the create flow, it's more important that those options exist after creation because you're more likely to start simple and evolve to complexity"?

smarterclayton commented:
Yes.


> On Apr 6, 2015, at 8:41 AM, Ben Parees <notifications@github.com> wrote:
> 
> @smarterclayton :
> 
> Saying "complexity in a create flow is less important" could be interpreted as "it's ok to be super complex in the create flow, it's not a problem, so let's make it super complex" or "it's not necessary to include complexity in the create flow". I'm assuming you meant the latter, so let me attempt to paraphrase: "it's less critical we provide all the possible options during the create flow, it's more important that those options exist after creation because you're more likely to start simple and evolve to complexity"?
> 
> —
> Reply to this email directly or view it on GitHub.
>

jwforres commented on assets/app/scripts/controllers/create.js:
```diff
@@ -0,0 +1,62 @@
+    DataService.list("templates", $scope, function(templates) {
+      $scope.projectTemplates = templates.by("metadata.name");
+      templatesByTag();
+      console.log("project templates", $scope.projectTemplates);
```
Switch the console.logs in this file to use Logger

jwforres replied:
Still need to switch logs in this file to be Logger

jwforres commented on assets/app/scripts/controllers/create/createFromImage.js:
```diff
@@ -0,0 +1,223 @@
+        replicas: 1
+      };
+
+      DataService.get("imageRepositories", scope.imageName, scope, {namespace: scope.namespace})
```
do you need to retrieve both the imageRepo and the imageRepoTags ? 

Also you should switch these to use imageStreams and imageStreamTags (would need to rebase on master from friday)

jcantrill commented:
@jwforres @bparees  Updated PR to reduce the routing options, utilize 'imageStreams', use logging service, style changes.  This pull will create a single service (headless or not based on exposed ports) and an optional unsecure route based on our discussion with @smarterclayton and @danmcp from this morning.

jcantrill commented:
@jwforres @csrwng @bparees  Looking for a final review to get this merged...

liggitt commented on assets/app/scripts/controllers/create/createFromImage.js:
```diff
@@ -0,0 +1,186 @@
+      };
+
+      DataService.get("imageStreams", scope.imageName, scope, {namespace: scope.namespace})
+        .then(function(imageRepo){
```
keep the `.then` on the same line

liggitt commented on assets/app/scripts/controllers/create/createFromImage.js:
```diff
@@ -0,0 +1,186 @@
+          scope.imageRepo = imageRepo;
+          var imageName = scope.imageTag;
+          DataService.get("imageStreamTags", imageRepo.metadata.name + ":" + imageName, {namespace: scope.namespace})
+            .then(function(image){
```
ditto

liggitt commented on assets/app/scripts/controllers/create/createFromImage.js:
```diff
@@ -0,0 +1,186 @@
+      });
+      
+      ifResourcesDontExist(resources, $scope.namespace, $scope)
+        .then(
```
ditto

csrwng commented on assets/app/scripts/services/applicationGenerator.js:
```diff
@@ -0,0 +1,267 @@
+
+  .service("ApplicationGenerator", function(){
+    var osApiVersion = "v1beta1";
+    var k8eApiVersion = "v1beta3";
```
s/k8e/k8s/

csrwng commented on assets/app/scripts/services/applicationGenerator.js:
```diff
@@ -0,0 +1,267 @@
+          {
+            containerPort: parseInt(parts[0]), 
+            name: input.name + "-" + parts[1] + "-" + parts[0],
+            protocol: parts[1]
```
a port spec may not have a protocol part ("80" and "80/tcp" are both valid). The default protocol should be "tcp"

csrwng commented on assets/app/scripts/services/applicationGenerator.js:
```diff
@@ -0,0 +1,267 @@
+        deploymentConfig: scope._generateDeploymentConfig_(input, imageSpec, ports, input.labels),
+        service: scope._generateService_(input, input.name, scope._getFirstPort_(ports))
+      };
+      resources.route = scope._generateRoute_(input, resources.service.metadata.name, resources.service.metadata.name);
```
Is there a reason route is not just part of the resources hash above it?

jcantrill replied:
@csrwng One issue here is its named wrong, should be input.name as one arg...but it is called out separately here because I am using the service name as input to the function.

csrwng commented on assets/app/scripts/json/imagerepository.json:
```diff
@@ -0,0 +1,21 @@
+{
```
What's the purpose of this file?

jcantrill replied:
@csrwng deleting...i think this is an artifact of an earlier WIP

jcantrill commented:
@jwforres @liggitt @csrwng @bparees updated based on @csrwng and @liggitt comments, squashed and rebased

csrwng commented:
@jcantrill what about the comment above about the port spec? You should default to "tcp" if a second part doesn't exist

csrwng commented:
@jcantrill also did you address the issue with the error notifications when checking existence of objects?

liggitt commented on assets/app/scripts/services/applicationGenerator.js:
```diff
@@ -0,0 +1,267 @@
+        var parts = key.split("/");
+        ports.push(
+          {
+            containerPort: parseInt(parts[0]), 
```
watch our for NaN here... just make sure we never show that in the UI

jwforres commented on .gitignore:
```diff
@@ -2,6 +2,7 @@
 /openshift.local.*
 /.project
 /.vagrant
+/assets/nbproject
```
what is this for?  if its something only you are going to have locally, then lets not check this in

jcantrill replied:
its an artifact left by using the netbeans IDE.  Isnt it easier to update the .gitignore file then having to exclude it everytime we make changes.  presumably im the only one using the ide..but it doesnt cost us anything

jwforres commented on assets/app/index.html:
```diff
@@ -120,8 +120,12 @@
+        <script src="scripts/services/applicationGenerator.js"></script>
         <script src="scripts/services/login.js"></script>
         <script src="scripts/services/logout.js"></script>
+        <script src="scripts/services/labelFilter.js"></script>
```
this should not be in index.html anymore, bad merge maybe?

jcantrill replied:
@jwforres removing

jwforres commented on assets/app/scripts/filters/yesNo.js:
```diff
@@ -0,0 +1,8 @@
+"use strict";
+
+angular.module('openshiftConsole')
+  .filter('yesNo', function() {
```
instead of this being in its own file lets just put it into filters/util.js

jwforres commented on assets/app/scripts/filters/httpHttps.js:
```diff
@@ -0,0 +1,8 @@
+"use strict";
+
+angular.module('openshiftConsole')
+  .filter('httpHttps', function() {
```
same here, lets just move this to filters/util.js

jwforres commented on assets/app/scripts/filters/httpHttps.js:
```diff
@@ -0,0 +1,8 @@
+
+angular.module('openshiftConsole')
+  .filter('httpHttps', function() {
+    return function(input) {
```
input seems like the wrong name for this parameter, change it to something like isSecure

jwforres commented on assets/app/scripts/filters/valuesIn.js:
```diff
@@ -0,0 +1,22 @@
+"use strict";
+
+angular.module("openshiftConsole")
+  .filter("valuesIn", function(){
```
move to filters/util.js

jwforres commented on assets/app/scripts/filters/valuesNotIn.js:
```diff
@@ -0,0 +1,22 @@
+   * @param {String} keys    A comma delimited string of keys to evaluate against
+   * @returns {Hash} A filtered set where the keys of those not in keys
+   */
+  .filter("valuesNotIn", function(){
```
move to filters/util.js

jwforres commented on assets/app/scripts/controllers/catalog/images.js:
```diff
@@ -0,0 +1,74 @@
+              $scope.builders.push(image);
+            }
+          });
+          console.log("builders", $scope.builders);          
```
switch to Logger

jcantrill replied:
@jwforres updating

jwforres commented on assets/app/scripts/controllers/catalog/images.js:
```diff
@@ -0,0 +1,74 @@
+      $scope.projectImageRepos = imageRepos.by("metadata.name");
+      imagesForRepos($scope.projectImageRepos, $scope);
+
+      console.log("project image repos", $scope.projectImageRepos);
```
Logger

jcantrill replied:
@jwforres updating

jwforres commented on assets/app/scripts/controllers/catalog/images.js:
```diff
@@ -0,0 +1,74 @@
+      $scope.openshiftImageRepos = imageRepos.by("metadata.name");
+      imagesForRepos($scope.openshiftImageRepos, {namespace: "openshift"});
+
+      console.log("openshift image repos", $scope.openshiftImageRepos);
```
Logger

jwforres commented on assets/app/scripts/controllers/catalog/images.js:
```diff
@@ -0,0 +1,74 @@
+        }
+      });
+
+      console.log("templatesByTag", $scope.templatesByTag);
```
Logger

jwforres commented on assets/app/scripts/controllers/create/createFromImage.js:
```diff
@@ -0,0 +1,183 @@
+        scope.name = NameGenerator.suggestFromSourceUrl(scope.buildConfig.sourceUrl);
+    }
+
+    initAndValidate($scope);
```
why is this done as a function if it is only ever called once?

liggitt commented on assets/app/scripts/services/navigate.js:
```diff
@@ -0,0 +1,16 @@
+        $location.url(redirect);
+      },
+      toProjectOverview: function(projectName){
+        $location.path("project/" + projectName + "/overview");
```
wrap projectName in `encodeURIComponent(...)`

liggitt commented on assets/app/scripts/services/navigate.js:
```diff
@@ -0,0 +1,16 @@
+angular.module("openshiftConsole")
+  .service("Navigate", function($location){
+    return {
+      toErrorPage: function(message) {
```
optional second param for error_code

jwforres commented on assets/app/scripts/controllers/newfromtemplate.js:
```diff
@@ -133,7 +133,7 @@ angular.module('openshiftConsole')
             return d.promise;
           });
-          $location.path("project/" + $scope.projectName + "/overview");
+          Navigate.toProjectOverview($scope.projectName);
```
would rather we had a service that knew how to generate URLs and return them, there will be cases where we need the URLs and won't necessarily be wanting to navigate right then.

jwforres commented on assets/app/styles/_core.less:
```diff
@@ -4,6 +4,7 @@
 
 html, body {
   height: 100%;
+  word-wrap: break-word;
```
@sg00dwin is looking at another alternative for this

sg00dwin replied:
@jcantrill 
This was included in my earlier pull request as a legacy carry over from V2, but let's remove it and apply that on a per case basis.

jcantrill replied:
@sg00dwin  Removed.  Will be commited in next push

jwforres commented on assets/app/views/catalog/_image.html:
```diff
@@ -0,0 +1,49 @@
+              <p ng-if="imageRepo | annotation : (imageTag + '.description')">{{imageRepo | annotation : (imageTag + '.description')}}</p>
+              <p>
+                <div ng-if="imageRepo | provider">
+                  <label style="margin-right: 5px;">Provider:</label>{{imageRepo | provider}}
```
this should be pulling the provider annotation based on the tag, `{{imageRepo | annotation : (imageTag + '.provider')}}`  and if it doesnt exist fallback to showing the namespace the imageRepo is in

jcantrill replied:
@jwforres Updated...Also fixed bug where the model did not pop if the tag has dots (e.g. v0.4)

jwforres commented on assets/app/scripts/controllers/create/createFromImage.js:
```diff
@@ -0,0 +1,183 @@
+      var resources = [];
+      angular.forEach(resourceMap, function(value, key){
+        if(value !== null){
+          Logger.debug("Generated resource definition:" + angular.toJson(value));
```
you should be able to do
`Logger.debug("Generated resource definition:", value)`  and then you'll get the inspectable object in the console.  Unless you were specifically wanting to dump the raw json to see what was going to be posted.

jcantrill replied:
Updating...its the same.

jwforres commented on assets/app/scripts/directives/oscFileInput.js:
```diff
@@ -0,0 +1,31 @@
+      restrict: 'E',
+      scope: {
+        name: "@",
+        model: "=",
```
lets call this what it is, which is the fileContents (or something similar)

jcantrill replied:
@jwforres this is supposed to reflect that this is our version of '<input type=file >'.  Do we really want to call it 'filecontent'?

liggitt replied:
I feel like this comment is missing something

liggitt commented on assets/app/scripts/directives/oscFileInput.js:
```diff
@@ -0,0 +1,31 @@
+        $(element).change(function(){
+          var file = $('input[type=file]',this)[0].files[0];
+          scope.fileName = file.name;
+          scope.$apply(function(){
```
pretty sure the scope.$apply should be inside the onloadend handler

jwforres replied:
agreed, since onloadend is async from the rest of the block

liggitt replied:
should be
```
reader.onloadend = function() {
  scope.$apply(function() {
    ...
  });
};
```

jcantrill replied:
fixed

jwforres commented on assets/app/scripts/directives/oscFormSection.js:
```diff
@@ -0,0 +1,24 @@
+      },
+      templateUrl: "views/directives/osc-form-section.html",
+      link: function(scope, element, attrs){
+        if(!attrs.editText) attrs.editText="Edit";
```
why are you pulling editText and expand off of the attrs instead of the scope?

jcantrill replied:
@jwforres here we are simply trying to default a value if they are not defined.  I believe they are always undefined on the scope

jwforres commented on assets/app/scripts/directives/oscResourceName.js:
```diff
@@ -0,0 +1,35 @@
+"use strict";
+
+angular.module("openshiftConsole")
+  .directive("oscResourceName", function(){
```
is this directive only adding a validator?  if so might be nice to call it oscResourceNameValidator, then it is clear in the templates where you are using it that is what it is adding

jcantrill replied:
Updated to oscResourceNameValidator

jwforres commented on assets/app/scripts/services/applicationGenerator.js:
```diff
@@ -0,0 +1,270 @@
+
+  .service("ApplicationGenerator", function(){
+    var osApiVersion = "v1beta1";
+    var k8sApiVersion = "v1beta3";
```
these are still hardcoded, they need to be returned as a method from DataService, something like DataService.osAPIVersion and DataService.k8sAPIVersion.  I don't want the console talking to the API server on potentially different version numbers because we forget to update it somewhere.

jcantrill replied:
Updated to be in DataService

jwforres commented on assets/app/scripts/services/applicationGenerator.js:
```diff
@@ -0,0 +1,270 @@
+    var service = {};
+    var scope = service;
+    
+    scope._generateSecret_ = function(){
```
style nit, drop the trailing _ on these method names.  For private methods we just use the single leading _

jwforres replied:
why are all these private methods on the scope object? unless the template needs to be able to access them (in which case it is strange that they are private functions) do `var _generateSecret = function` instead

jcantrill replied:
The only reason they are accessible is to simply testing.  It allows us 
the ability to call them individually.

On 04/08/2015 03:15 PM, Jessica Forrester wrote:
> In assets/app/scripts/services/applicationGenerator.js
> <https://github.com/openshift/origin/pull/1221#discussion_r28003247>:
>
>> @@ -0,0 +1,270 @@
>> +"use strict";
>> +
>> +angular.module("openshiftConsole")
>> +
>> +  .service("ApplicationGenerator", function(){
>> +    var osApiVersion = "v1beta1";
>> +    var k8sApiVersion = "v1beta3";
>> +
>> +    var service = {};
>> +    var scope = service;
>> +
>> +    scope._generateSecret_ = function(){
>
> why are all these private methods on the scope object? unless the
> template needs to be able to access them (in which case it is strange
> that they are private functions) do |var _generateSecret = function| instead
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/1221/files#r28003247>.
>

-- 
Jeff Cantrill
Senior Software Engineer, Red Hat Engineering
Red Hat
Office: 703-748-4420 | 866-546-8970 ext. 8162420
jcantril@redhat.com
http://www.redhat.com

jwforres commented on assets/app/scripts/services/applicationGenerator.js:
```diff
@@ -0,0 +1,270 @@
+    * @return {integer} The port/protocol pair of the lowest conainer port
+    */
+    scope._getFirstPort_ = function(ports){
+      var first = "None";
```
are we actually passing "None" back to the API somewhere? otherwise it seems like this should return null when there is not a port

jcantrill replied:
The value of "None" is used when specifying a headless service.

On 04/08/2015 03:18 PM, Jessica Forrester wrote:
> In assets/app/scripts/services/applicationGenerator.js
> <https://github.com/openshift/origin/pull/1221#discussion_r28003419>:
>
>> +        //http://stackoverflow.com/questions/105034/create-guid-uuid-in-javascript
>> +        function s4() {
>> +          return Math.floor((1 + Math.random()) * 0x10000)
>> +              .toString(16)
>> +              .substring(1);
>> +          }
>> +        return s4()+s4()+s4()+s4();
>> +      };
>> +
>> +    /**
>> +    * Find the 'first' port of exposed ports.
>> +    * @param            ports  list of ports (e.g {containerPort: 80, protocol: "tcp"})
>> +    * @return {integer} The port/protocol pair of the lowest conainer port
>> +    */
>> +    scope._getFirstPort_ = function(ports){
>> +      var first = "None";
>
> are we actually passing "None" back to the API somewhere? otherwise it
> seems like this should return null when there is not a port
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/1221/files#r28003419>.
>

-- 
Jeff Cantrill
Senior Software Engineer, Red Hat Engineering
Red Hat
Office: 703-748-4420 | 866-546-8970 ext. 8162420
jcantril@redhat.com
http://www.redhat.com

jwforres commented on assets/app/scripts/services/applicationGenerator.js:
```diff
@@ -0,0 +1,270 @@
+     * @returns Hash of resource definitions
+     */
+    scope.generate = function(input){
+      //map ports to k8e structure
```
fix comment, k8s

jwforres commented on assets/app/scripts/services/applicationGenerator.js:
```diff
@@ -0,0 +1,270 @@
+          name: input.name, 
+          tag: "latest",
+          toString: function(){
+            return this.name + ":" + this.tag;
```
above you reference input.name, here you reference this.name. Is there a difference?  If not lets be consistent, one way or the other.

jcantrill replied:
Isn't the use of 'this' more correct since 'this' will refer to the hash 
that is imageSpec?

On 04/08/2015 03:24 PM, Jessica Forrester wrote:
> In assets/app/scripts/services/applicationGenerator.js
> <https://github.com/openshift/origin/pull/1221#discussion_r28003944>:
>
>> +            name: input.name + "-" + parts[1] + "-" + parts[0],
>> +            protocol: parts[1]
>> +          });
>> +      });
>> +
>> +      //augment labels
>> +      input.labels.name = input.name;
>> +      input.labels.generatedby = "OpenShiftWebConsole";
>> +
>> +      var imageSpec;
>> +      if(input.buildConfig.sourceUrl !== null){
>> +        imageSpec = {
>> +          name: input.name,
>> +          tag: "latest",
>> +          toString: function(){
>> +            return this.name + ":" + this.tag;
>
> above you reference input.name, here you reference this.name. Is there a
> difference? If not lets be consistent, one way or the other.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/1221/files#r28003944>.
>

-- 
Jeff Cantrill
Senior Software Engineer, Red Hat Engineering
Red Hat
Office: 703-748-4420 | 866-546-8970 ext. 8162420
jcantril@redhat.com
http://www.redhat.com

jwforres replied:
Only if the function is called with the correct scope.  You can always 
change the scope a function is called in :)

If this is an issue of name and tag potentially being changed later, 
then I understand why you would want to use this.  If name and tag arent 
going to be changed by anything, then i'd say inject the actual values.

On 04/09/2015 02:12 PM, Jeff Cantrill wrote:
>
> In assets/app/scripts/services/applicationGenerator.js 
> <https://github.com/openshift/origin/pull/1221#discussion_r28087291>:
>
> > +            name: input.name + "-" + parts[1] + "-" + parts[0],
> > +            protocol: parts[1]
> > +          });
> > +      });
> > +
> > +      //augment labels
> > +      input.labels.name = input.name;
> > +      input.labels.generatedby = "OpenShiftWebConsole";
> > +
> > +      var imageSpec;
> > +      if(input.buildConfig.sourceUrl !== null){
> > +        imageSpec = {
> > +          name: input.name,
> > +          tag: "latest",
> > +          toString: function(){
> > +            return this.name + ":" + this.tag;
> Isn't the use of 'this' more correct since 'this' will refer to the 
> hash that is imageSpec?
> … <#>
> On 04/08/2015 03:24 PM, Jessica Forrester wrote: In 
> assets/app/scripts/services/applicationGenerator.js 
> <https://github.com/openshift/origin/pull/1221#discussion_r28003944>: 
> > + name: input.name + "-" + parts[1] + "-" + parts[0], > + protocol: 
> parts[1] > + }); > + }); > + > + //augment labels > + 
> input.labels.name = input.name; > + input.labels.generatedby = 
> "OpenShiftWebConsole"; > + > + var imageSpec; > + 
> if(input.buildConfig.sourceUrl !== null){ > + imageSpec = { > + name: 
> input.name, > + tag: "latest", > + toString: function(){ > + return 
> this.name + ":" + this.tag; above you reference input.name, here you 
> reference this.name. Is there a difference? If not lets be consistent, 
> one way or the other. — Reply to this email directly or view it on 
> GitHub <https://github.com/openshift/origin/pull/1221/files#r28003944>.
> -- Jeff Cantrill Senior Software Engineer, Red Hat Engineering Red Hat 
> Office: 703-748-4420 | 866-546-8970 ext. 8162420 jcantril@redhat.com 
> http://www.redhat.com
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/openshift/origin/pull/1221/files#r28087291>.
>

jwforres commented on assets/app/scripts/services/applicationGenerator.js:
```diff
@@ -0,0 +1,270 @@
+      return resources;
+    };
+    
+    scope._generateRoute_ = function(input, name, serviceName){
```
I feel like all these generate functions should really be abstracted out. Some kind of service that generates JSON for a type based on a set of inputs.  The second we want to create/update individual components we are going to have to duplicate this.  

Significant set of changes at this point, we can do that as a refactor later.

jcantrill replied:
I would say that's a strong maybe.  Update operations would allow us to 
modify the existing json so I'm not sure it applies.  Also, creating 
individual resources I would think we might want to offer a user more 
options then the simplified version of 'Generate Application'.

On 04/08/2015 03:48 PM, Jessica Forrester wrote:
> In assets/app/scripts/services/applicationGenerator.js
> <https://github.com/openshift/origin/pull/1221#discussion_r28005808>:
>
>> +            return this.name + ":" + this.tag;
>> +          }
>> +        };
>> +      }
>> +
>> +      var resources = {
>> +        imageRepo: scope._generateImageRepo_(input),
>> +        buildConfig: scope._generateBuildConfig_(input, imageSpec, input.labels),
>> +        deploymentConfig: scope._generateDeploymentConfig_(input, imageSpec, ports, input.labels),
>> +        service: scope._generateService_(input, input.name, scope._getFirstPort_(ports))
>> +      };
>> +      resources.route = scope._generateRoute_(input, input.name, resources.service.metadata.name);
>> +      return resources;
>> +    };
>> +
>> +    scope._generateRoute_ = function(input, name, serviceName){
>
> I feel like all these generate functions should really be abstracted
> out. Some kind of service that generates JSON for a type based on a set
> of inputs. The second we want to create/update individual components we
> are going to have to duplicate this.
>
> Significant set of changes at this point, we can do that as a refactor
> later.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/1221/files#r28005808>.
>

-- 
Jeff Cantrill
Senior Software Engineer, Red Hat Engineering
Red Hat
Office: 703-748-4420 | 866-546-8970 ext. 8162420
jcantril@redhat.com
http://www.redhat.com

jwforres replied:
Sure an independent create may allow more operations, but the "model" 
service or whatever it would be called should in theory handle varied 
levels of input, when stuff isn't specified it would take defaults.

I just don't want us duplicating JSON building all over the place if it 
isn't necessary, makes dealing with API model changes more troublesome.

On 04/09/2015 02:15 PM, Jeff Cantrill wrote:
>
> In assets/app/scripts/services/applicationGenerator.js 
> <https://github.com/openshift/origin/pull/1221#discussion_r28087565>:
>
> > +            return this.name + ":" + this.tag;
> > +          }
> > +        };
> > +      }
> > +
> > +      var resources = {
> > +        imageRepo: scope._generateImageRepo_(input),
> > +        buildConfig: scope._generateBuildConfig_(input, imageSpec, input.labels),
> > +        deploymentConfig: scope._generateDeploymentConfig_(input, imageSpec, ports, input.labels),
> > +        service: scope._generateService_(input, input.name, scope._getFirstPort_(ports))
> > +      };
> > +      resources.route = scope._generateRoute_(input, input.name, resources.service.metadata.name);
> > +      return resources;
> > +    };
> > +
> > +    scope._generateRoute_ = function(input, name, serviceName){
> I would say that's a strong maybe. Update operations would allow us to 
> modify the existing json so I'm not sure it applies. Also, creating 
> individual resources I would think we might want to offer a user more 
> options then the simplified version of 'Generate Application'.
> … <#>
> On 04/08/2015 03:48 PM, Jessica Forrester wrote: In 
> assets/app/scripts/services/applicationGenerator.js 
> <https://github.com/openshift/origin/pull/1221#discussion_r28005808>: 
> > + return this.name + ":" + this.tag; > + } > + }; > + } > + > + var 
> resources = { > + imageRepo: scope._generateImageRepo_(input), > + 
> buildConfig: scope._generateBuildConfig_(input, imageSpec, 
> input.labels), > + deploymentConfig: 
> scope._generateDeploymentConfig_(input, imageSpec, ports, 
> input.labels), > + service: scope._generateService_(input, input.name, 
> scope._getFirstPort_(ports)) > + }; > + resources.route = 
> scope._generateRoute_(input, input.name, 
> resources.service.metadata.name); > + return resources; > + }; > + > + 
> scope._generateRoute_ = function(input, name, serviceName){ I feel 
> like all these generate functions should really be abstracted out. 
> Some kind of service that generates JSON for a type based on a set of 
> inputs. The second we want to create/update individual components we 
> are going to have to duplicate this. Significant set of changes at 
> this point, we can do that as a refactor later. — Reply to this email 
> directly or view it on GitHub 
> <https://github.com/openshift/origin/pull/1221/files#r28005808>.
> -- Jeff Cantrill Senior Software Engineer, Red Hat Engineering Red Hat 
> Office: 703-748-4420 | 866-546-8970 ext. 8162420 jcantril@redhat.com 
> http://www.redhat.com
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/openshift/origin/pull/1221/files#r28087565>.
>

jwforres commented on assets/app/views/create/fromimage.html:
```diff
@@ -0,0 +1,227 @@
+                </div>
+                <div class="has-error" ng-show="nameTaken">
+                  <span class="help-block">This name is not unique within the scope of the project. Please choose a different one.
+                  <p>Resources within OpenShift require a unique name within scope of your project.  This name is already taken.</p>
```
Having both sentences seems excessive.
How about just saying:
"This name is already in use within the project. Please choose a different name."

jwforres commented on assets/app/views/create/fromimage.html:
```diff
@@ -0,0 +1,227 @@
+                about-title="Routing"
+                about="Routing is a way to make your application publicly visible.  You may only be
+                    able to access your application by it's IP address otherwise if allowed by
+                    the system administrator."
```
"Routing is a way to make your application publicly visible. Otherwise you may only be able to access your application by it's IP address, if allowed by the system administrator."

jwforres commented on assets/app/views/create/fromimage.html:
```diff
@@ -0,0 +1,227 @@
+                      <input type="checkbox" ng-model="routing" ng-checked="routing">
+                        Create a route to the application
+                        <span class="help action-inline">
+                          <i class="pficon pficon-help" data-toggle="tooltip" data-placement="right" data-original-title="Additional routes can be added by using the command-line interface with alternative configuration options (e.g. security, TLS termination, etc)."></i>
```
"Additional routes with alternative configuration options (security, TLS termination, etc) can be added using the osc command."

jwforres commented on assets/app/views/create/fromimage.html:
```diff
@@ -0,0 +1,227 @@
+                  </div>
+                  <h3>Environment Variables <span class="help action-inline">
+                    <a href data-toggle="tooltip" data-placement="right" 
+                       data-original-title="Environment variables are used to configure and pass information to images.">
```
s/images/running containers/ ?

jwforres commented on assets/app/views/create/fromimage.html:
```diff
@@ -0,0 +1,227 @@
+              <osc-form-section 
+                header="Build Configuration"
+                about-title="Build Configuration"
+                about="A build configuration describes how to build your deployable image which includes
```
Either  ", which includes" or ". This includes"

jwforres commented on assets/app/views/create/fromimage.html:
```diff
@@ -0,0 +1,227 @@
+                header="Build Configuration"
+                about-title="Build Configuration"
+                about="A build configuration describes how to build your deployable image which includes
+                  your source, the image upon which to base the final image, and when to launch new builds."
```
instead of "the image upon which to base the final image"  maybe "the base image to build with"

jwforres commented on assets/app/views/create/fromimage.html:
```diff
@@ -0,0 +1,227 @@
+                    <input type="number" class="input-number" min="0" name="replicas" ng-model="scaling.replicas" required ng-pattern="/^\-?\d+$/">
+                  </label>
+                  <div class="has-error" ng-show="form.replicas.$invalid">
+                    <span class="help-block">Replicas must be an integer value greater than or equal 0</span>
```
"or equal to 0"

jwforres commented on assets/app/views/create/fromimage.html:
```diff
@@ -0,0 +1,227 @@
+              <div class="buttons gutter-top-bottom gutter-top-bottom-2x">
+              <div class="alert alert-info">
+                <span class="pficon pficon-info"></span>
+                <label>Note:</label> After creation, these settings can only be modified through the cli.
```
instead of cli lets say "osc command"

jwforres commented on assets/app/views/directives/osc-key-values.html:
```diff
@@ -0,0 +1,77 @@
+      <ul class="list-unstyled label-list">
+        <li ng-repeat="(key,value) in entries | valuesIn:readonlyKeys">
+          <span class="key truncate">{{ key}}</span>
+          <span class="value truncate">{{ value}}</span>
```
clean up the extra space after the {{ on both of these, and the same ones below

jwforres commented on examples/image-repositories/image-repositories.json:
```diff
@@ -6,9 +6,16 @@
       "kind": "ImageRepository",
       "metadata": {
-        "name": "ruby-20-centos7"
+        "name": "ruby-20-centos7",
```
the changes to this file are no longer valid, @ncdc did you submit a PR with the correct version already? if not we can put those changes here.  @jcantrill  if andy's changes are already merged just remove the changes to this file from your PR.

liggitt commented on assets/app/scripts/directives/oscFileInput.js:
```diff
@@ -0,0 +1,31 @@
+              scope.model = reader.result;
+            };
+            reader.onerror = function(e){
+              Logger.error(e);
```
not sure what to do here... we can't just leave the file selected and not tell them about the error... maybe flip `scope.supportsFileUpload` to false so they can paste in the file? If we do that, we should also notify them that there was an error reading the file or they'll be really confused

jwforres commented:
[merge]

bparees commented:
woo!  awesome work!

pweil- commented:
:trophy:

jcantrill commented:
[merge]
